vms/platformvm/state/subnet_only_validator.go

```diff
@@ -128,3 +210,184 @@ func putSubnetOnlyValidator(db database.KeyValueWriter, vdr SubnetOnlyValidator)
 func deleteSubnetOnlyValidator(db database.KeyValueDeleter, validationID ids.ID) error {
 	return db.Delete(validationID[:])
 }
+
+type subnetOnlyValidatorsDiff struct {
+	numAddedActive      int               // May be negative
+	modifiedTotalWeight map[ids.ID]uint64 // subnetID -> totalWeight
+	modified            map[ids.ID]SubnetOnlyValidator
+	modifiedHasNodeIDs  map[subnetIDNodeID]bool
+	active              *btree.BTreeG[SubnetOnlyValidator]
+}
+
+func newSubnetOnlyValidatorsDiff() *subnetOnlyValidatorsDiff {
+	return &subnetOnlyValidatorsDiff{
+		modifiedTotalWeight: make(map[ids.ID]uint64),
+		modified:            make(map[ids.ID]SubnetOnlyValidator),
+		modifiedHasNodeIDs:  make(map[subnetIDNodeID]bool),
+		active:              btree.NewG(defaultTreeDegree, SubnetOnlyValidator.Less),
+	}
+}
+
+// getActiveSubnetOnlyValidatorsIterator takes in the parent iterator, removes
+// all modified validators, and then adds all modified active validators.
+func (d *subnetOnlyValidatorsDiff) getActiveSubnetOnlyValidatorsIterator(parentIterator iterator.Iterator[SubnetOnlyValidator]) iterator.Iterator[SubnetOnlyValidator] {
+	return iterator.Merge(
+		SubnetOnlyValidator.Less,
+		iterator.Filter(parentIterator, func(sov SubnetOnlyValidator) bool {
+			_, ok := d.modified[sov.ValidationID]
+			return ok
+		}),
+		iterator.FromTree(d.active),
+	)
+}
+
+func (d *subnetOnlyValidatorsDiff) hasSubnetOnlyValidator(subnetID ids.ID, nodeID ids.NodeID) (bool, bool) {
+	subnetIDNodeID := subnetIDNodeID{
+		subnetID: subnetID,
+		nodeID:   nodeID,
+	}
+	has, modified := d.modifiedHasNodeIDs[subnetIDNodeID]
+	return has, modified
+}
+
+func (d *subnetOnlyValidatorsDiff) putSubnetOnlyValidator(state Chain, sov SubnetOnlyValidator) error {
```

Comment:
`putSubnetOnlyValidator` is essentially directly called by `state.PutSubnetOnlyValidator` and `diff.PutSubnetOnlyValidator`. So I suppose the question is why those functions are tested rather than this one directly?

It feels much more complex to me to need to implement a test-double for `Chain` rather than using either `Diff` or `State`, which is what the existing tests are already doing.

(I also generally prefer testing exported functions over un-exported functions, as there is a more clear separation between the caller and the callee).